File: AdventOfCode2024/Go/Day13/Day13.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type coord struct {
	x float64
	y float64
}

type machine struct {
	a coord
	b coord
	p coord
}

func parse(filename string) []machine {
	file, _ := os.Open(filename)
	fscanner := bufio.NewScanner(file)

	machines := []machine{}
	newMachine := machine{}
	for fscanner.Scan() {
		line := fscanner.Text()

		if line == "" {
			machines = append(machines, newMachine)
			newMachine = machine{}
		}

		if strings.Split(line, ":")[0] == "Button A" {
			aX, _ := strconv.Atoi(strings.Split(strings.Split(strings.Split(line, ":")[1], ",")[0], "+")[1])
			aY, _ := strconv.Atoi(strings.Split(strings.Split(strings.Split(line, ":")[1], ",")[1], "+")[1])
			newMachine.a = coord{float64(aX), float64(aY)}
		}

		if strings.Split(line, ":")[0] == "Button B" {
			bX, _ := strconv.Atoi(strings.Split(strings.Split(strings.Split(line, ":")[1], ",")[0], "+")[1])
			bY, _ := strconv.Atoi(strings.Split(strings.Split(strings.Split(line, ":")[1], ",")[1], "+")[1])
			newMachine.b = coord{float64(bX), float64(bY)}
		}

		if strings.Split(line, ":")[0] == "Prize" {
			pX, _ := strconv.Atoi(strings.Split(strings.Split(strings.Split(line, ":")[1], ",")[0], "=")[1])
			pY, _ := strconv.Atoi(strings.Split(strings.Split(strings.Split(line, ":")[1], ",")[1], "=")[1])
			newMachine.p = coord{float64(pX), float64(pY)}
		}
	}
	machines = append(machines, newMachine)

	return machines
}

func tokensToWin(m machine, constant float64) int {

	px := m.p.x + (constant)
	py := m.p.y + (constant)

	// SO INT() IS THE SAME AS FLOOR() IN GO
	aPresses := math.Round((px - (m.b.x*m.p.y)/m.b.y) / (m.a.x - (m.b.x*m.a.y)/m.b.y))
	bPresses := math.Round((py - (m.a.y * aPresses)) / m.b.y)

	// cant have negative presses
	if aPresses < 0 || bPresses < 0 {
		return 0
	}

	// as we are using floats we need to check if the presses are whole numbers
	// easist way to do this is to assert if the eqs work
	if ((aPresses*m.a.x)+(bPresses*m.b.x) != px) || ((aPresses*m.a.y)+(bPresses*m.b.y) != py) {
		return 0
	}
	return int(aPresses*3 + bPresses)
}

func Part1(filename string, debug bool) int {
	machines := parse(filename)

	if debug {
		fmt.Println(machines)
	}
	ans := 0
	for _, m := range machines {
		ans += tokensToWin(m, 0)
	}

	return ans
}

func Part2(filename string, debug bool) int {
	machines := parse(filename)

	if debug {
		fmt.Println(machines)
	}
	ans := 0
	for _, m := range machines {
		ans += tokensToWin(m, 10000000000000)
	}

	return ans
}

func main() {
	fmt.Println("Advent-Of-Code 2024 - Day13")
	fmt.Printf("Tests : Answer to Part 1 = %v\n", Part1("tests.txt", false))
	fmt.Printf("Inputs: Answer to Part 1 = %v\n", Part1("inputs.txt", false))
	fmt.Println("had to use python for part 2 :(")
}

File: AdventOfCode2024/Python/Day13/Day13.py
import numpy as np

class machine:
    def __init__(self):
        self.ax = 0
        self.ay = 0
        self.bx = 0
        self.by = 0
        self.px = 0
        self.py = 0

with open("inputs.txt") as f:
        data =((f.read()).replace(" ","")).split("\n")

machines = []
newMachine = machine()
for line in data:
    if line == "":
        machines.append(newMachine)
        newMachine = machine()
        continue

    if "ButtonA" in line:
        newMachine.ax, newMachine.ay = (int(line.split("+")[1].split(",")[0]), int(line.split("+")[-1]))
    
    if "ButtonB" in line:
        newMachine.bx, newMachine.by = (int(line.split("+")[1].split(",")[0]), int(line.split("+")[-1]))
    
    if "Prize" in line:
        newMachine.px, newMachine.py = (int(line.split("=")[1].split(",")[0]), int(line.split("=")[-1]))
    
machines.append(newMachine)

def Solve(constant):

    ans = 0
    for m in machines:

        px = m.px + constant
        py = m.py + constant

        a = np.array([[m.ax, m.bx], [m.ay, m.by]])
        b = np.array([px, py])
        s = np.linalg.solve(a, b)
        aPresses, bPresses = np.round(s[0]), np.round(s[1])

        if aPresses < 0 or bPresses < 0:
            continue

        if ((aPresses*m.ax)+(bPresses*m.bx) != px) or ((aPresses*m.ay)+(bPresses*m.by) != py):
            continue
        
        ans += 3*aPresses + bPresses

    return ans

print("Advent-Of-Code 2024 - Day13\n")
print(f'Inputs: Answer to Part 1 = {Solve(0)}')
print(f'Inputs: Answer to Part 2 = {Solve(10000000000000)}')
